test/linalg/matmul.jl

```diff
@@ -389,3 +389,33 @@ let
         @test_throws DimensionMismatch A_mul_B!(full43, full43, tri44)
     end
 end
+
+# #18218
+module TestPR18218
+  using Base.Test
+  import Base.*, Base.+, Base.zero
+  immutable TypeA
+    x::Int
+  end
```

Comment:
Not that it matters too much, but you can save some space by changing, e.g. `TypeA` by `Int`, so you only need two types more for the test.